### DISCUSSION TITLE
Generating Relationships with automap_base

### DISCUSSION THREAD

**Discussion by gabishop88** [sqlalchemy/sqlalchemy#11388] [Usage Questions] [answered]
My issue may be an overcomplicated Database Schema, but I am trying to generate the table classes with automap_base, since it would be quite painful to write them all out by hand. As a general outline of my project, I am using sqlalchemy and geoalchemy2 with a postgresql (PostGIS) backend and FastAPI. My db has three main schemas, `public`, `frontend`, and `reference`. There are FKs that reference tables between schemas. Here is what I have tried with automap_base so far, based on [this](https://github.com/sqlalchemy/sqlalchemy/discussions/7102) discussion:

```python
Base = automap_base()
Base.metadata.reflect(
    engine,
    schema='reference',
    extend_existing=True,
    autoload_replace=False
)
Base.metadata.reflect(
    engine,
    schema='frontend',
    extend_existing=True,
    autoload_replace=False
)
Base.metadata.reflect(
    engine,
    schema='public',
    only=['farm', 'field', 'subfield', 'field_usage', 'crop_usage', 'field_action'],
    extend_existing=True,
    autoload_replace=False
)
Base.prepare()
```

At one point, I had something simple, like `Base.prepare(autoload_with=engine)`, since I was only using tables in public for the api at the time. I am trying to access more tables outside of that now. Before, I had Farm.field_collection as the relation between 'farm' and 'field' tables. My main issue is that now when I use prepare(), there are no onetomany relationships (like tablename_collection) The only one that was generated was a collection for a table that I don't even want a collection for since it is a manytomany relation, but it generated in the wrong direction for how I want to use it. On the other hand, every single table in the database is in Base.classes, just not the relationships I want.

Here are a couple of examples of what I want to happen, and what I got.

I have tables: public.field_usage, public.crop_usage, and reference.crop. crop_usage is just a m2m relation between field_usage and crop. I would like to have FieldUsage.crop_collection based on crop_usage, but instead I get crop.field_usage_collection, which is useless to me.

As far as I can tell, every single 1 to M relation generated a scalar relationship, e.g. the public.field table has Field.farm_id as a FK to Farm.farm_id. This generated Field.farm, but not Farm.field_collection. This is pretty much the situation for all of the FKs with the exception of the m2m relation example above. 

As a note, I am referring to the db tables in snake case (as they are in the db) and the class names in CamelCase, since I also passed a function for camel case into prepare. 

If there is anything else I can provide to help, let me know. I can also alter the database to change field names, etc. if that would resolve the issue. 

I want to mention the the results also seem a little unpredictable. Sometimes, I will change either nothing or something small or unrelated, and suddently different relationships are generated. I don't have an example right now, but I can follow up with one if I can reproduce the effect.

Sorry this is so long, but here is the result of me printing the name of each table after prepare() along with each property in \_\_mapper__.iterate_properties. Notice that since I generated this after I wrote the above example, Field.farm no longer exists, as for some reason, these generate in a different order almost every time, and sometimes generated relationships are just left out. I am beginning to suspect that this has something to do with the order the tables are read in?

```python
Class: YieldUnit (comes from schema: reference)
         name
Class: Farm (comes from schema: None)
         farm_id
         farm_name
         description
Class: UserAccess (comes from schema: frontend)
         appuser                 [Generated]
         farm                    [Generated]
         user_id
         farm_id
         can_read
         can_edit
         can_delete
Class: FieldAction (comes from schema: public)
         producttype             [Generated]
         actiontype              [Generated]
         fieldusage              [Generated]
         yieldunit               [Generated]
         applicationrateunit     [Generated]
         action_id
         usage_id
         type
         date
         duration
         prod_type
         applied_product
         application_rate
         app_rate_unit
         product_manufacturer
         harvested_product
         yield_per_acre
         yield_per_acre_unit
         yield_moisture_percent
         biomass_removed_percent
         maintenance_task
         notes
Class: UsageType (comes from schema: reference)
         name
         description
Class: Crop (comes from schema: reference)
         fieldusage_collection   [Generated]
         crop_id
         common_name
         latin_name
         cultivar
         description
Class: Subfield (comes from schema: public)
         fieldtype               [Generated]
         field                   [Generated]
         subfield_id
         parent_field_id
         type
         subfield_name
         organic_since
         active
         notes
         geom
Class: YieldConversion (comes from schema: reference)
         yieldunit               [Generated]
         unit
         product
         lbs_per_acre_conv
Class: FieldUsage (comes from schema: public)
         subfield                [Generated]
         field                   [Generated]
         usagetype               [Generated]
         usage_id
         field_id
         subfield_id
         type
         name
         start
         end
         notes
Class: ApplicationRateUnit (comes from schema: reference)
         name
         lbs_per_acre_conv
Class: AppUser (comes from schema: frontend)
         user_id
         email
         full_name
         hashed_password
         is_active
         is_superuser
Class: ActionType (comes from schema: reference)
         name
         description
Class: FieldType (comes from schema: reference)
         name
         description
Class: MaintenanceTask (comes from schema: reference)
         name
         description
Class: ProductType (comes from schema: reference)
         name
         decription
Class: Field (comes from schema: public)
         fieldtype               [Generated]
         field_id
         farm_id
         type
         field_name
         group_label
         organic_since
         active
         notes
         geom

```

The last thing I wanted to mention is about `sqlacodegen`. I tried to use this tool instead (as I  have been struggling with this problem for a few days now) and have even attempted to use tools like peewee and pwiz. sqlacodegen I got to work, but generated the classes with similar relationship problems. I will probably just manually edit all of the sqlacodegen classes if I cannot figure this out, but I am going full sunken cost at this point.

**Comment by CaselIT** [comment, marked as answer]
Hi,

In postgresql, unless you change the default `search_path`, you should not use the schema `public` explicitly in sqlalchemy. Try changing the last reflect call to remove `schema='public'`.
More info here https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#remote-schema-table-introspection-and-postgresql-search-path

**Reply by gabishop88** [reply]
Thank you for your reply. This seems to have mostly fixed the issue, I believe. Such a simple thing. 

The only question I have is that if I try to access Base.classes.Farm.field_collection, I get an error saying it doesn't exist. However, when I query something that uses Farm.field_collection, it seems to work. Do you have any guesses as to why?

**Reply by gabishop88** [reply]
Just in case, here is what I mean:

database.py (where I have `Base = automap_base()`):
```python
print(Base.classes.Farm.field_collection)
```
> AttributeError: type object 'Farm' has no attribute 'field_collection'


crud.py:
```python
farm_query = db.query(Farm).outerjoin(fields_sq).outerjoin(subfields_sq).options(
    contains_eager(Farm.field_collection, alias=fields_sq).
    contains_eager(Field.subfield_collection, alias=subfields_sq)
).filter(Farm.farm_id.in_(farm_ids))
```
> No Problems

Also, I am (obviously) quite new to SQLAlchemy, so if there is something that's not great about how I am going about this, feel free to lmk.

**Reply by CaselIT** [reply]
That seems because the class is not fully configured yet. Sqlalchemy uses lazy init that automatically triggers in some cases, for example when using the class on a select.

You can manually trigger it using https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.configure_mappers. does calling it before accessing the relationship help?

**Reply by gabishop88** [reply]
Yes! That makes a lot of sense. I had a feeling before that some stuff was just not set up, but didn't know how to cause it to do so. Thank you for the help!
